### PR TITLE
Support for dynamic codec adapter id - Part 1

### PR DESCRIPTION
**Pull request by dbaluta**
This patch series only move around in order to prepare for sending the codec_adapter ID at runtime from Linux kernel side.

We move all Cadence API calls from init() to prepare() because codec_id will not be available at init(). Earliest case where we could get the codec_id from Linux kernel is at set_params(). the call to prepare() is done later so at the point the codec_id will be know.

### CODE REVIEW

File: src/audio/codec_adapter/codec/cadence.c [hunk 1]
```diff
@@ -681,16 +680,6 @@ cadence_codec_set_configuration(struct processing_module *mod, uint32_t config_i
 	    md->state < MODULE_INITIALIZED)
 		return 0;
 
-	/* whole configuration received, apply it now */
```
Comment (ranj063) [inline src/audio/codec_adapter/codec/cadence.c:655]: @dbaluta if we dont apply the config right away here, the check just above is redundant right?

Comment (dbaluta) [reply src/audio/codec_adapter/codec/cadence.c:655]: @ranj063 true.

File: src/audio/codec_adapter/codec/cadence.c [hunk 2]
```diff
@@ -667,31 +667,11 @@ cadence_codec_set_configuration(struct processing_module *mod, uint32_t config_i
...
-		comp_err(dev, "cadence_codec_set_configuration(): error %x: runtime config apply failed",
-			 ret);
-		return ret;
-	}
-
-	comp_dbg(dev, "cadence_codec_set_configuration(): config applied");
+	comp_dbg(mod->dev, "cadence_codec_set_configuration()");
 
-	return 0;
+	return module_set_configuration(mod, config_id, pos, data_offset_size, fragment,
```
Comment (ranj063) [inline src/audio/codec_adapter/codec/cadence.c:673]: would it better to remove cadence_codec_set_configuration() and set ..set_configuration = model_set_configuration directly?

Comment (dbaluta) [reply src/audio/codec_adapter/codec/cadence.c:673]: That would be an option too. I will add it in a separate patch to be more obvious.

**Review by lgirdwood — APPROVED** [review]

**Review by paulstelian97 — APPROVED** [review]

**Comment by dbaluta** [comment]
SOFCI TEST

**Comment by gkbldcig** [comment]
Can one of the admins verify this patch?

**Comment by dbaluta** [comment]
SOFCI TEST

File: src/audio/codec_adapter/codec/cadence.c [hunk 3]
```diff
@@ -94,13 +94,59 @@ static struct cadence_api cadence_api_table[] = {
...
+	}
+	/* Get codec object size */
+	API_CALL(cd, XA_API_CMD_GET_API_SIZE, 0, &obj_size, ret);
+	if (ret != LIB_NO_ERROR) {
+		comp_err(dev, "cadence_codec_init() error %x: failed to get lib object size",
+			 ret);
+		return ret;
+	}
+	/* Allocate space for codec object */
+	cd->self = rballoc(0, SOF_MEM_CAPS_RAM, obj_size);
```
[the remark this answers is not in the document]
Comment (lyakh) [reply src/audio/codec_adapter/codec/cadence.c:156]: As far as I understand generic memory allocation functions in SOF are `rmalloc()` and `rzalloc()` with `rballoc()` reserved for audio buffer allocations. I know, that `rballoc()` is also used across the code base for other purposes, which may or may not be justified. With XTOS the differences are roughly, that `rballoc()` allocates from the largest memory pool so it has a higher chance of success. With Zephyr all allocations are performed from the same heap, so there there are hardly any practical differences. So to me it's like - `rmalloc()` is the default, `rzalloc()` if you need to 0-initialise, and `rballoc()` for large allocations or if you have any other very special reasons to use it.

[the remark this answers is not in the document]
Comment (dbaluta) [reply src/audio/codec_adapter/codec/cadence.c:156]: @lyakh this PR just move code around. I would prefer to take care of your  comments in a separate PR. 

Cc: @ranj063

[the remark this answers is not in the document]
Comment (lgirdwood) [reply src/audio/codec_adapter/codec/cadence.c:156]: ack - lets deal with functional changes later.

**Comment by lgirdwood** [comment]
@kkarask do you have an ETA for CI readyness to test ?

**Comment by dbaluta** [comment]
@kkarask I would also help if you provide me with a way to run your tests. It looks like you are injecting IPC messages into FW.

File: src/audio/codec_adapter/codec/cadence.c [hunk 4]
```diff
@@ -667,31 +667,11 @@ cadence_codec_set_configuration(struct processing_module *mod, uint32_t config_i
...
-	if (ret < 0)
-		return ret;
-
-	/* return if more fragments are expected or if the module is not prepared */
-	if ((pos != MODULE_CFG_FRAGMENT_LAST && pos != MODULE_CFG_FRAGMENT_SINGLE) ||
-	    md->state < MODULE_INITIALIZED)
-		return 0;
 
-	/* whole configuration received, apply it now */
-	ret = cadence_codec_apply_config(dev);
```
[the remark this answers is not in the document]
Comment (dbaluta) [reply src/audio/codec_adapter/codec/cadence.c:656]: Fixed with  your comments. Hope this is clear now.

File: src/audio/codec_adapter/codec/cadence.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (lyakh) [reply src/audio/codec_adapter/codec/cadence.c:156]: @dbaluta sure, works for me

File: src/audio/codec_adapter/codec/cadence.c [hunk 5]
```diff
@@ -94,13 +94,59 @@ static struct cadence_api cadence_api_table[] = {
...
+		comp_err(dev, "cadence_codec_init() error %x: failed to get lib name",
+			 ret);
+		return ret;
+	}
+	/* Get codec object size */
+	API_CALL(cd, XA_API_CMD_GET_API_SIZE, 0, &obj_size, ret);
+	if (ret != LIB_NO_ERROR) {
+		comp_err(dev, "cadence_codec_init() error %x: failed to get lib object size",
+			 ret);
+		return ret;
```
[the remark this answers is not in the document]
Comment (lyakh) [reply src/audio/codec_adapter/codec/cadence.c:153]: @dbaluta do you plan submitting a PR for this in the near future or should I file a bug for this to make sure it isn't lost? This seems a severe enough bug to me.

[the remark this answers is not in the document]
Comment (dbaluta) [reply src/audio/codec_adapter/codec/cadence.c:153]: Please fill a bug so we can track it. There is definitely a lot of improvements we can do wrt codec adapter.

[the remark this answers is not in the document]
Comment (lyakh) [reply src/audio/codec_adapter/codec/cadence.c:153]: #5658

**Comment by dbaluta** [comment]
@kkarask @ranj063 @lgirdwood this should work fine now. After talking with Filip I finally got an understanding of how the failing test expects codec_adapter to behave when setting runtime params.

**Comment by dbaluta** [comment]
SOFCI TEST

**Pull request merged**